libbeat/outputs/redis/redis.go

```diff
@@ -118,19 +124,57 @@ func makeRedis(
 	}
 
 	clients := make([]outputs.NetworkClient, len(hosts))
-	for i, host := range hosts {
-		enc, err := codec.CreateEncoder(beat, config.Codec)
+	for i, h := range hosts {
+		if parts := strings.SplitN(h, "://", 2); len(parts) != 2 {
+			h = fmt.Sprintf("%s://%s", redisScheme, h)
+		}
+
+		hostUrl, err := url.Parse(h)
+		if err != nil {
+			return outputs.Fail(err)
+		}
+
+		if hostUrl.Host == "" {
+			return outputs.Fail(fmt.Errorf("invalid redis url host %s", hostUrl.Host))
+		}
+
+		if hostUrl.Scheme != redisScheme && hostUrl.Scheme != tlsRedisScheme {
+			return outputs.Fail(fmt.Errorf("invalid redis url scheme %s", hostUrl.Scheme))
+		}
+
+		if hostUrl.Scheme == tlsRedisScheme && transp.TLS == nil {
+			transp.TLS = &transport.TLSConfig{}
+		}
```

Comment:
Nice, this should work. If `transp.TLS != nil`, then TLS will always be used. I just noticed that `transp` is a shared variable. We should create `transp` within the loop, such that each client (host) has it's own set of settings and can individually enable/disable TLS.

We might want to add an `if else` here to disable TLS in case `redis` is used. This way users can configure TLS, but still use URLS with and without encryption.

I think this should do the trick:

```
hasScheme := true
parts := strings.SplitN(h, "://", 2); len(parts) != 2 {
  h = fmt.Sprintf("%s://%s", redisScheme, h)
  hasScheme = false
}


switch hostUrl.Scheme {
case redisScheme:
  if hasScheme {
    transp.TLS = nil // disable TLS if user explicitely set `redis` scheme
  }
case tlsRedisScheme:
  if transp.TLS == nil {
    transp.TLS = &transport.TLSConfig{} // enable with system default if TLS was not configured
  }
}
```